### PR TITLE
[GBNews]Add new extractor for GB News TV channel

### LINKED ISSUE #29341: Add support for gbnews.uk
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.06.06**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Video: https://www.gbnews.uk/shows/andrew-neils-message-to-companies-choosing-to-boycott-gb-news/106889
- Live: https://www.gbnews.uk/watchlive

## Description

GB News uses simplestreamcdn for videos. Despite mention of widevine and fairplay in the js, they don't seem to serve drm-protected content.

I was able to make a shell script to negotiate with the simplestreamcdn api using jq and httpie. Might help with adding support. I'm unsure if sending "heartbeats" is required to keep the livestream working

```bash
function gbnews-dl () {
    local SSMP="https://mm-dev.simplestream.com/ssmp/api.php"
    local PLATFORM="safari" # Stored as a global "platform". Likely could effect if drm is served

    local article=$(http -b "$1")
    local element=$(echo "$article" | xmllint --html --xpath "//*[@class='simplestream']" - 2>/dev/null)

    if [ -z "$element" ]; then
        echo "gbnews-dl: Can't find simplestream element!" 1>&2
        return 1
    fi

    local datasetId=$(echo "$element" | xmllint --xpath "string(/div/@data-id)" -)
    local datasetEnv=$(echo "$element" | xmllint --xpath "string(/div/@data-env)" -)
    local datasetKey=$(echo "$element" | xmllint --xpath "string(/div/@data-key)" -)
    local datasetType=$(echo "$element" | xmllint --xpath "string(/div/@data-type)" -)
    local datasetUvid=$(echo "$element" | xmllint --xpath "string(/div/@data-uvid)" -)
    local datasetToken=$(echo "$element" | xmllint --xpath "string(/div/@data-token)" -)
    local datasetExpiry=$(echo "$element" | xmllint --xpath "string(/div/@data-expiry)" -)

    if [ "$datasetType" = "vod" ]; then
        datasetType="show"
    fi

    echo "gbnews-dl: Attempt lookup for $datasetId" 1>&2

    local sSMeta=$(http -b "$SSMP" "id==$datasetId" "env==$datasetEnv")
    echo "$sSMeta" | jq 2>&1 1>/dev/null
    if [ "$?" != 0 ]; then
        echo "gbnews-dl: Failed to parse metadata server response. Not valid JSON" 1>&2
        echo "$sSMeta" 1>&2
        return 1
    fi

    local sSOrigin=$(echo "$sSMeta" | jq -r '.response?.api_hostname')

    if [ -z "$sSOrigin" ]; then
        echo "gbnews-dl: No API hostname returned. Blocked by server?" 1>&2
        echo "$sSMeta" | jq '.' 1>&2
        return 1
    fi

    local streamMeta=$(http -b "$sSOrigin/api/$datasetType/stream/$datasetUvid" \
                       "key==$datasetKey" "platform==$PLATFORM" "Token:$datasetToken" \
                       "Token-Expiry:$datasetExpiry" "Uvid:$datasetUvid")
    echo "$streamMeta" | jq 2>&1 1>/dev/null
    if [ "$?" != 0 ]; then
        echo "gbnews-dl: Failed to parse negotiation server response. Not valid JSON" 1>&2
        echo "$streamMeta" 1>&2
        return 1
    fi

    local streamUsesDRM=$(echo "$streamMeta" | jq -r '.response?.drm != null')

    if [ "$streamUsesDRM" = "true" ]; then
        echo "gbnews-dl: Stream is requesting DRM playback. Unsupported" 1>&2
        echo "$streamMeta" | jq '.' 1>&2
        return 1
    fi

    # TODO: Videos play fine without messing with heartbeats however sending them might be required to keep
    #       the livestream url working. Needs more looking into
    local streamUrl=$(echo "$streamMeta" | jq -r '.response.stream')

    if [ -z "$streamUrl" ]; then
        echo "gbnews-dl: Successful negotiation but no stream URL provided!" 1>&2
        echo "$streamMeta" | jq '.' 1>&2
        return 1
    fi

    echo "$streamUrl"
}
```

### PR DESCRIPTION
**Pull request by dirkf**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Issue #29341 identified the need for an extractor for the new GB News channel from the UK, and demonstrated extraction using a shell script.

This PR adds an extractor for the two types of GB News URLs (clips, live) shown in the issue, using the procedure from the issue.

The live stream appears to operate OK with a shell command line piping yt-dl output to `mpv`, but hasn't been extensively exercised.

Closes #29341.

### CODE REVIEW

**Comment by kakolisgay** [comment]
Small issue in the URL regex, videos on gbnews are not exclusive to `/show`, the can also appear as other types, as can be seen at https://www.gbnews.uk/news/sedgefield-company-creates-revolutionary-covid-detection-machine/109879
Other than that, seems good 👍

**Comment by dirkf** [comment]
Apparently any text (let's say alphanumeric and _) can go in place of `shows` and the link redirects to the preferred version.

File: youtube_dl/extractor/gbnews.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    _TESTS = [{
+        'url': 'https://www.gbnews.uk/shows/andrew-neils-message-to-companies-choosing-to-boycott-gb-news/106889',
+        'info_dict': {
+            'id': '106889',
+            'ext': 'mp4',
+            'title': "Andrew Neil's message to companies choosing to boycott GB News",
+            'description': 'md5:b281f5d22fd6d5eda64a4e3ba771b351',
+        },
+    },
+    ]
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:31]: New tests:
```suggestion
    }, {
        'url': 'https://www.gbnews.com/news/bbc-claudine-gay-harvard-university-antisemitism-row',
        'info_dict': {
            'id': '52264136',
            'display_id': 'bbc-claudine-gay-harvard-university-antisemitism-row',
            'ext': 'mp4',
            'title': 'BBC deletes post after furious backlash over headline downplaying antisemitism',
            'description': 'The post was criticised by former employers of the broadcaster',
        },
    }, {
        'url': 'https://www.gbnews.uk/watchlive',
        'info_dict': {
            'id': '1069',
            'display_id': 'watchlive',
            'ext': 'mp4',
            'title': 'GB News Live',
            'is_live': True,
        },
        'params': {
            'skip_download': 'm3u8',
        },
    }]
```

File: youtube_dl/extractor/gbnews.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:34]: Support deeper URL paths:
```suggestion
        display_id = self._match_id(url).split('/')[-1]
```

File: youtube_dl/extractor/gbnews.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            data-expiry="1624300052"
+            data-uvid="37900558"
+            data-poster="https://thumbnails.simplestreamcdn.com/gbnews/ondemand/37900558.jpg?width=700&"
+            data-npaw="false"
+            data-env="production">
+        '''
+        # exception if no match
+        video_data = self._search_regex(
+            r'<\s*div\s[^>]*class\s*=\s*([\'"])simplestream\1[^>]*>',
+            webpage, "video data", group=0)
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:54]: Update pattern:
```suggestion
        video_data = self._search_regex(
            r'(<div\s[^>]*\bclass\s*=\s*(\'|")(?!.*sidebar\b)simplestream(?:\s[\s\w$-]*)?\2[^>]*>)',
            webpage, 'video data')
```

File: youtube_dl/extractor/gbnews.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+        # print(video_data)
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:56]:
```suggestion
```

File: youtube_dl/extractor/gbnews.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        video_data = extract_attributes(video_data)
+        ss_id = try_get(video_data, lambda x: x['data-id'])
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:58]: `video_data` is known to be a `dict`.
```suggestion
        ss_id = video_data.get('data-id')
```

File: youtube_dl/extractor/gbnews.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        json_data = self._download_json(
+            self._SSMP_URL, display_id,
+            note='Downloading Simplestream JSON metadata',
+            errnote='Unable to download Simplestream JSON metadata',
+            query={
+                'id': ss_id,
+                'env': video_data.get('data-env'),
+            })
+
+        meta_url = try_get(json_data, lambda x: x['response']['api_hostname'], compat_str)
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:72]: Rework, providing default `env` from site JS:
```suggestion
        json_data = self._download_json(
            self._SSMP_URL, display_id,
            note='Downloading Simplestream JSON metadata',
            errnote='Unable to download Simplestream JSON metadata',
            query={
                'id': ss_id,
                'env': video_data.get('data-env', 'production'),
            }, fatal=False)

        meta_url = traverse_obj(json_data, ('response', 'api_hostname'))
```

File: youtube_dl/extractor/gbnews.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        if not meta_url:
+            raise ExtractorError('No API host found')
+
+        uvid = video_data.get('data-uvid')
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:76]: This is the actual video ID:
```suggestion
        uvid = video_data['data-uvid']
```

File: youtube_dl/extractor/gbnews.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            },
+            headers={
+                'Token': video_data.get('data-token'),
+                'Token-Expiry': video_data.get('data-expiry'),
+                'Uvid': uvid,
+            })
+
+        stream_url = try_get(stream_data, lambda x: x['response']['stream'], compat_str)
+        if not stream_url:
+            raise ExtractorError('No stream data')
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:94]: Rework with `uvid`:
```suggestion
        stream_data = self._download_json(
            '%s/api/%s/stream/%s' % (meta_url, 'show' if dtype == 'vod' else dtype, uvid),
            uvid,
            query={
                'key': video_data.get('data-key'),
                'platform': self._PLATFORM,
            },
            headers={
                'Token': video_data.get('data-token'),
                'Token-Expiry': video_data.get('data-expiry'),
                'Uvid': uvid,
            }, fatal=False)

        stream_url = traverse_obj(stream_data, (
            'response', 'stream', T(url_or_none)))
        if not stream_url:
            raise ExtractorError('No stream data/URL')
```

File: youtube_dl/extractor/gbnews.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+        # now known to be a dict
+        stream_data = stream_data['response']
+        drm = stream_data.get('drm')
+        if drm:
+            raise ExtractorError(
+                'Stream is requesting DRM (%s) playback: unsupported' % drm,
+                expected=True)
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:102]: Now we can do this:
```suggestion
            self.report_drm(uvid)
```

File: youtube_dl/extractor/gbnews.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+        formats = []
+        formats.extend(
+            self._extract_m3u8_formats(stream_url, display_id, ext='mp4', fatal=False))
+
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:107]: Simplify:
```suggestion
        formats = self._extract_m3u8_formats(
            stream_url, uvid, ext='mp4', entry_protocol='m3u8_native',
            fatal=False)
```

File: youtube_dl/extractor/gbnews.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        # exception if no formats
+        self._sort_formats(formats)
+
+        # no 'title' attribute seen, but if it comes ...
+        title = stream_data.get('title') or self._og_search_title(webpage)
+
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:113]:
```suggestion
```

File: youtube_dl/extractor/gbnews.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        return {
+            'id': display_id,
+            'title': title,
+            'description': self._og_search_description(webpage, default=None),
+            'thumbnail': video_data.get('data-poster') or None,
+            'formats': formats,
+            'is_live': 'Live' in self.IE_NAME,
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:120]:
```suggestion
            'id': uvid,
            'display_id': display_id,
            'title': (traverse_obj(stream_data, ('title', T(txt_or_none)))
                      or self._og_search_title(webpage, default=None)
                      or display_id.replace('-', ' ').capitalize()),
            'description': self._og_search_description(webpage, default=None),
            'thumbnail': (traverse_obj(video_data, ('data-poster', T(url_or_none)))
                          or self._og_search_thumbnail(webpage)),
            'formats': formats,
            'is_live': (dtype == 'live') or None,
```

File: youtube_dl/extractor/gbnews.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        data-type="live"
+        data-key="3Li3Nt2Qs8Ct3Xq9Fi5Uy0Mb2Bj0Qs"
+        data-token="d10b3ea37f6ce539ffd1ce2f6ce5fe35"
+        data-expiry="1624984755"
+        data-uvid="1069"
+        data-poster=""
+        data-npaw="false"
+        data-env="production">
+    </div>
+    '''
```
Comment (dirkf) [inline youtube_dl/extractor/gbnews.py:152]: No longer needed.
```suggestion
```

**Review by dirkf — COMMENTED** [review]
Update for 2024

**Pull request merged**